### ISSUE TITLE
Looping code doesn't terminate

### ISSUE THREAD

**Issue by elliotweiser** [mbj/mutant#327]
I had difficulty pointing `mutant` at code with loops.  The line responsible for decrementing a timeout was being removed as part of mutation.  It would be helpful if I could see the mutations that were being made before the test rather than after, precisely for this reason.  This could be written to a file so as to not flood STDOUT.

In any case, the non-terminating behavior is fixable.  I added the following code to my `spec_helper.rb`:

``` ruby
def wait_for_test(timeout: 10, signal: 'SIGTERM')
  pid = fork do
    yield
  end
  sleep(timeout)
  result, status = Process.waitpid2(pid, Process::WNOHANG)
  if result.nil?
    Process.kill(signal, pid)
    fail 'Test did not complete in time...'
  end
  fail 'Tests failed' if status.exitstatus != 0
end
```

Now I can run my expect statements within a `wait_for_test { }` block.

**Comment by elliotweiser** [comment]
Unfortunately, this exact code will not work everywhere-- `fork` is not supported on all platforms (Windows/JRuby).  I'm considering making this code available in a separate gem that can function with the Windows OS and a JVM.

**Comment by mbj** [comment]
@elliotweiser Can you show me your loop construct that runs into non terminating mutations?

Most often its possible to refactor such loops into semantically equivalent versions that are not mutated into non-terminating ones.

Or adding a very "implementation specific expectation" that kills that mutation before that infinite loop gets mutated in.

**Comment by elliotweiser** [comment]
My apologies, I should have posted the code/tests in this issue as well.

My code:

```ruby
module MyModule
  module MyClass
    def poll_for_success
      timeout = DEFAULTS[:timeout]
      while timeout > 0
        return true if check_for_success
        timeout = sleep_for_interval_and_decrement_timeout(timeout)
      end
      false
    end
  end
end
```

My RSpec test:

```ruby
describe MyModule::MyClass do
  context '#poll_for_success' do
    let(:actual) do
      subject.poll_for_success
    end
    it 'should call :sleep_for_interval_and_decrement_timeout twice and return false' do
      MyModule::DEFAULTS[:timeout] = 0.02
      MyModule::DEFAULTS[:interval] = 0.01
      expect(subject)
        .to receive(:sleep_for_interval_and_decrement_timeout)
        .with(0.02).and_return(0.01).ordered
      expect(subject)
        .to receive(:sleep_for_interval_and_decrement_timeout)
        .with(0.01).and_return(0).ordered
      expect(actual).to be false
    end
  end
end
```

The solution I presented in my first comment I believe addresses the second option you described: 

> Or adding a very "implementation specific expectation" that kills that mutation before that infinite loop > gets mutated in.

This occurs at test-level rather than mutant level, namely by failing the test if it takes too long.

Just to be clear, I don't believe that this behavior represents a bug in mutant.  It can be easily explained, even if the exact mutation is not known.  That said, it would be a big help if the guesswork were taken out of the process i.e. by exposing the mutations _before_ the test run.  

To summarize, here is my reasoning for opening this issue:

- Inform the developers/users of `mutant` of the infinite-runtime behavior.
- Feature request: Log each mutation in real-time (as it is being used), just in case things don't go as expected.

**Comment by elliotweiser** [comment]
Also, thank you for developing this tool!  I can only speak for myself at the moment, but it has proven to be an invaluable component of my development and CI processes.  Before `mutant`, even with 100% code coverage, it turns out my test coverage was super shoddy.  I'm really excited to get some of the rails devs I work with to start using it!

**Comment by kbrock** [comment]
I added a timeout for my tests to work around this.

I found it tricky to work around the infinite loop issue.
I would introduce and remove that issues every so often.

**Comment by tjchambers** [comment]
@elliotweiser I would say your comment - I appreciate this tool immensely. Not only does it critique my test coverage (in a good way), but it also makes me re-evaluate my coding techniques in a way that prevents bugs. It is truly a key tool in my toolbox.

**Comment by mbj** [comment]
There had been long discussions about infinite loops in https://github.com/mbj/mutant/issues/87 and https://github.com/mbj/mutant/issues/81.

I'm busy with non OSS stuff, I'll do a long response, preferably in the form of a blog post before closing this issue.

**Comment by elliotweiser** [comment]
@tjchambers well stated.

@mbj My preference is to avoid using the `Rspec.configure` block because it applies too broadly. The tests for my functions containing while loops may very well take less time than my other tests. Configuring a sufficient timeout setting would add unnecessary overhead.  For this reason, I think the freedom to specify a timeout for a _particular_ test as it is needed is very valuable.

**Comment by mbj** [comment]
@elliotweiser You can also add the timeout just for the subject that need them?

```
RSpec.describe TheSubjectThatProducesInfiniteLoops do
  around do |example|
    Timeout.timeout(1, &example)
  end
end
```

This solves the problem for me. And I'm not convinced that mutant side timeouts are any good:

* Mutated code should be killable via editing in the mutation maually and just running the tests
* If a mutation is killed just from mutant side timeouts: It cannot be manually killed anymore.

For now I'll close the issue, as I think there is nothing to gain via expanding mutant into killing mutations via timeouts on the mutant side.

**Comment by mvz** [comment]
@mbj please reconsider adding a timeout option on the mutant side.

**Comment by mvz** [comment]
Practically speaking, I would go for kills, since they would be test failures in CI.

**Comment by mvz** [comment]
I'm trying to understand where you would report these in any case, since killed mutants are currently not reported at all?

**Comment by mbj** [comment]
@mvz I'd simply start to add another "detected mutation result" state. And let the user (via a config file) decide on how the detected effect of the mutation (which will be one of: none, test failure, integration failure, external timeout) be handled. I'll likely also offer presets for those mappings.

I'd recommend to handle external timeouts as alives, but you could just map that state to killed.

**Comment by mvz** [comment]
@mbj ok, clear. That sounds like a good approach.

**Comment by mbj** [comment]
So this is the default mapping as I imagine it in the future:

| Type | Detection | Result | Abort |
| ----- | ------- | ------- | ---- | 
| Noop |  Tests passed | Dead | No |
| Noop | Tests failed | Alive | Yes |
| Noop | Isolation Error | Alive | Yes |
| Noop | Timeout | Alive | Yes |
| Evil | Tests passed | Alive | No |
| Evil | Tests failed | Dead | No |
| Evil | Isolation Error | Alive | No |
| Evil | Timeout | Alive | No |

This mapping will be configurable by the user. But I'll strongly advice against doing so :P

**Comment by mvz** [comment]
Well, I would definitely count timeout as Dead, since I don't want to add a timeout to my own test suite myself (mainly for debugging reasons).

**Comment by mbj** [comment]
This is an unsharp line, I'll write up why its a bad idea. And can let you do what you need.

**Comment by mbj** [comment]
Mutant with #1104 just got support for external timeouts.

Followups will allow to declare these as covered. But as external timeouts are now implemented I'm closing this issue.

**Comment by mbj** [comment]
Link to documentation for this issue: https://github.com/mbj/mutant/blob/master/docs/configuration.md#mutation_timeout
